packages/flutter/lib/src/widgets/localizations.dart

```diff
@@ -0,0 +1,465 @@
+
+  // Some of input.values were asynchronous futures. Wait for them.
+  return Future.wait<V>(outputFutures.values).then<Map<K, V>>((List<V> values) {
+    final List<K> keys = outputFutures.keys.toList();
+    for (int i = 0; i < keys.length; i++)
+      output[keys[i]] = values[i];
+    return output;
+  });
+}
+
+/// Defines all of an application's resources in terms of `allLoaders`:
+/// a map from a [Type] to a _load_ function that creates an instance of
+/// that type for a specific locale.
+///
+/// It's assumed that the that each loader will create an object that
+/// contains a collection or related resources (typically
+/// defined by one method per resource). The objects will be retrieved
+/// by [Type] with [resourcesFor].
+///
+/// Each loader returns a Future because it may need to do work
+/// asynchronously. For example localized resources might be loaded from
+/// the nextwork. The [load] method returns a Future that completes
+/// when all of the loaders have completed.
+///
+/// See also:
+///
+///  * [WidgetsApp.localizationsDelegate] and [MaterialApp.localizationsDelegate],
+///    which enable configuring the app's LocalizationsLoader.
+class DefaultLocalizationsDelegate extends LocalizationsDelegate {
+  DefaultLocalizationsDelegate(this.allLoaders) {
+    assert(allLoaders != null);
+  }
+
+  /// The [LocalizationsLoader]s in this map define all of the collections
+  /// of resources for a [Localizations] widget.
+  ///
+  /// The [load] method's Future completes when all of the loaders' load functions
+  /// have completed.
+  final Map<Type, LocalizationsLoader> allLoaders;
+
+  final Map<Locale, Map<Type, dynamic>> _localeToResources = <Locale, Map<Type, dynamic>>{};
+  final Set<Locale> _loading = new Set<Locale>();
+
+  @override
+  Future<Iterable<Type>> load(Locale locale) {
+    assert(locale != null);
+    assert(!_loading.contains(locale));
+
+    Iterable<Type> types;
+
+    _loading.add(locale);
+    final Future<Iterable<Type>> typesFuture = _loadMap<Type, dynamic>(
+      new Map<Type, Future<dynamic>>.fromIterables(
+        allLoaders.keys,
+        allLoaders.values.map<Future<dynamic>>((LocalizationsLoader loader) => loader(locale)),
+      ),
+    ).then<Iterable<Type>>((Map<Type, dynamic> resources) {
+      _loading.remove(locale);
+      _localeToResources[locale] = resources;
+      return types = resources.keys;
+    });
+
+    return types == null ? typesFuture : new SynchronousFuture<Iterable<Type>>(types);
+  }
+
+  @override
+  T resourcesFor<T>(Locale locale, Type type) {
+    assert(locale != null);
+    assert(type != null);
+    final Map<Type, dynamic> resources = _localeToResources[locale];
+    if (resources == null)
+      return null;
+    // TODO(hansmuller): assert(resources[type] is type) when
+    // https://github.com/dart-lang/sdk/issues/27680 has been resolved.
+    return resources[type];
+  }
+
+  @override
+  bool shouldReload(DefaultLocalizationsDelegate old) => false;
```

Comment:
That doesn't seem right. The normal case is that we'll be creating a new delegate every frame. Delegates are supposed to be short-lived.